src/main/java/com/minecolonies/api/crafting/BottleRecipe.java

```diff
@@ -0,0 +1,275 @@
+import net.minecraft.util.GsonHelper;
+import net.minecraft.world.inventory.CraftingContainer;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.*;
+import net.minecraft.world.level.ItemLike;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * A recipe to mix additional items into filled bottles.  Avoids duplicating the bottles.
+ */
+public class BottleRecipe extends ShapelessRecipe
+{
+    public BottleRecipe(@NotNull final ResourceLocation id,
+                        @NotNull final ItemStack output,
+                        @NotNull final NonNullList<Ingredient> inputs)
+    {
+        super(id, "", CraftingBookCategory.MISC, output, inputs);
+    }
+
+    @NotNull
+    @Override
+    public NonNullList<ItemStack> getRemainingItems(@NotNull final CraftingContainer container)
+    {
+        final NonNullList<ItemStack> remainingItems = super.getRemainingItems(container);
+        for (int i = 0; i < remainingItems.size(); ++i)
+        {
+            if (remainingItems.get(i).is(ModItems.large_empty_bottle))
+            {
+                remainingItems.set(i, ItemStack.EMPTY);
+            }
+        }
+        return remainingItems;
+    }
+
+    @NotNull
+    @Override
+    public RecipeSerializer<?> getSerializer()
+    {
+        return ModRecipeSerializer.BottleRecipeSerializer.get();
+    }
+
+    public static class Serializer implements RecipeSerializer<BottleRecipe>
+    {
+        @NotNull
+        @Override
+        public BottleRecipe fromJson(@NotNull final ResourceLocation id,
+                                     @NotNull final JsonObject json)
+        {
+            final ItemStack output = ShapedRecipe.itemStackFromJson(GsonHelper.getAsJsonObject(json, "result"));
+            final JsonArray array = GsonHelper.getAsJsonArray(json, "ingredients");
+            final NonNullList<Ingredient> inputs = NonNullList.create();
+            for (int i = 0; i < array.size(); ++i)
+            {
+                inputs.add(Ingredient.fromJson(array.get(i), false));
+            }
+
+            return new BottleRecipe(id, output, inputs);
+        }
+
+        @Nullable
+        @Override
+        public BottleRecipe fromNetwork(@NotNull final ResourceLocation id,
+                                        @NotNull final FriendlyByteBuf buf)
+        {
+            final int count = buf.readVarInt();
+            final NonNullList<Ingredient> inputs = NonNullList.withSize(count, Ingredient.EMPTY);
+            for (int i = 0; i < count; ++i)
+            {
+                inputs.set(i, Ingredient.fromNetwork(buf));
+            }
+            final ItemStack output = buf.readItem();
+
+            return new BottleRecipe(id, output, inputs);
+        }
+
+        @Override
+        public void toNetwork(@NotNull final FriendlyByteBuf buf,

```

Comment:
That feels like a lot of methods, do we need all of them? I thought this should be more of a straightforward copy of shapeless with only the remainingitems changing

Reply:
That's exactly what it is.  Though I lumped the `Builder` in there too, whereas with Shapeless it's in a separate file.

Pretty sure it will get shorter in 1.21, where a lot of this will be codecs instead, but I haven't looked closely at that yet.